ext/intl/spoofchecker/spoofchecker_main.c

```diff
@@ -167,3 +164,67 @@ PHP_METHOD(Spoofchecker, setRestrictionLevel)
 }
 /* }}} */
 #endif
+
+PHP_METHOD(Spoofchecker, setAllowedChars)
+{
+	zend_string *chars;
+	zend_long pattern_option = 0;
+	SPOOFCHECKER_METHOD_INIT_VARS;
+
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_STR(chars)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG(pattern_option)
+	ZEND_PARSE_PARAMETERS_END();
+	SPOOFCHECKER_METHOD_FETCH_OBJECT;
+
+	if (ZSTR_LEN(chars) > INT32_MAX) {
+		zend_argument_value_error(1, "is larger than " ZEND_LONG_FMT, INT32_MAX);
+		RETURN_THROWS();
+	}
+
+	/* uset_applyPattern requires to start with a regex range char */
+	bool is_pattern = ZSTR_VAL(chars)[0] == '[';
+	USet *set = uset_openEmpty();
+
+	if (!is_pattern && pattern_option) {
+		zend_argument_value_error(2, "pattern option set for non pattern settings");
+		RETURN_THROWS();
+	} else if (is_pattern && pattern_option) {
+		/* pattern is either USE_IGNORE_SPACE alone or in conjunction with the following flags (but mutually exclusive) */
+		if (pattern_option != USET_IGNORE_SPACE &&
+                    pattern_option != (USET_IGNORE_SPACE|USET_CASE_INSENSITIVE) &&
+                    pattern_option != (USET_IGNORE_SPACE|USET_ADD_CASE_MAPPINGS) &&
+		    pattern_option != (USET_IGNORE_SPACE|USET_SIMPLE_CASE_INSENSITIVE)) {
+			zend_argument_value_error(2, "invalid pattern option, must be (IGNORE_SPACE|(<none> or USET_CASE_INSENSITIVE or USET_ADD_CASE_MAPPINGS or USET_SIMPLE_CASE_INSENSITIVE))");
+			RETURN_THROWS();
+		}
+	}
+
+	if (is_pattern) {
+		uset_applyPattern(set, (UChar *)ZSTR_VAL(chars), (int32_t)ZSTR_LEN(chars), (uint32_t)pattern_option, SPOOFCHECKER_ERROR_CODE_P(co));
+		if (U_FAILURE(SPOOFCHECKER_ERROR_CODE(co))) {
+			zend_argument_value_error(1, "invalid pattern (%d) %s", SPOOFCHECKER_ERROR_CODE(co), u_errorName(SPOOFCHECKER_ERROR_CODE(co)));
+			uset_close(set);
+			RETURN_THROWS();
+		}
+	} else {
+		for (size_t i = 0; i < ZSTR_LEN(chars); i ++) {
+			UChar c = (UChar)ZSTR_VAL(chars)[i];
+			if (!uset_contains(set, c)) {
+				uset_add(set, c);
```

Comment:
How does this work for unicode codepoints? A unicode codepoints is more than one byte, but as far as I understand you're passing individual bytes to this function while the function accepts unicode codepoints.
So e.g. passing "\u{2460}" I would expect the unicode codepoint U+2460 to be added, but that corresponds to 3 bytes (in UTF-8).

Reply:
true, I may drop it .. also need to properly cast the input before passing to applyPattern.